### DISCUSSION TITLE
How to setup csp and global headers with tanstack start?

### DISCUSSION THREAD

**Discussion by discoverlance-com** [TanStack/router#3028] [Q&A]
This has been asked before in the discussion, https://github.com/TanStack/router/discussions/2476 but since there's no reply I am asking once more and with respect to tanstack start if anyone has an idea.

I want to configure some default http security headers like `x-frame-options` and such globally and also setup csp with a nonce if possible.

How can I approach this and has anyone done it before?

**Comment by PeterMylemans** [comment]
@discoverlance-com I'm using a vinxi middleware to do this.

Any feedback from the community is appreciated.

./app.config.ts
```typescript
import { defineConfig } from "@tanstack/start/config";
import tsConfigPaths from "vite-tsconfig-paths";

export default defineConfig({
  vite: {
    plugins: [
      tsConfigPaths({
        projects: ["./tsconfig.json"],
      }),
    ],
  },
  routers: {
    ssr: {
      middleware: "./app/middleware/ssr.ts",
    },
  },
});
```
./app/middleware/ssr.ts
```typescript
import { defineMiddleware, setResponseHeaders } from "vinxi/http";

export default defineMiddleware({
  onBeforeResponse: (event) => {
    setResponseHeaders(event, {
      "X-Frame-Options": "DENY",
      "X-Content-Type-Options": "nosniff",
      "Referrer-Policy": "strict-origin-when-cross-origin",
      "Permissions-Policy": "geolocation=(), camera=(), microphone=()",
      "Content-Security-Policy": "default-src 'self'; script-src 'self' 'unsafe-inline'"
    });
  },
});
```
Don't have a solution for nonce's yet but probably it is a post-processing somewhere in the SSR handler, provided a request attribute is set with the correct nonce (pretty similar to the nextjs solution):
 1. generate a nonce for an incoming SSR request and add the nonce in the request headers to store it in the _request memory_ (typically in a middleware)
 2. add the nonce to generated link and style refs once the HTML is generated (middleware or as feature of tanstack/start)
 3. add the nonce to the right spot in your CSP header (typically in a middleware)

**Comment by Enalmada** [comment]
I am hoping that tanstack start can be updated to support nonce like nextjs does:
https://nextjs.org/docs/app/building-your-application/configuring/content-security-policy

I think it may just be a change to Scripts component to take in a nonce parameter and render it in the script tags.

**Comment by kieranm** [comment]
I think the library really needs to prioritise proper support for this. It's very important in a production app to have a content security policy in place.

**Reply by viclafouch** [reply]
@tannerlinsley do you have any idea about this ?

**Comment by ebramanti** [comment]
Now that devinxi is in main, how can we handle CSP? Seems like https://github.com/TanStack/router/discussions/3028#discussioncomment-11691699 is no longer an option

**Comment by viclafouch** [comment]
On my side, it works with this : 

```ts
// src/server.ts
import type { RequestHandler } from '@tanstack/react-start/server'
import {
  createStartHandler,
  defaultStreamHandler
} from '@tanstack/react-start/server'
import { createRouter } from './router'

const handler = createStartHandler({
  createRouter
})(defaultStreamHandler)

const withHeaders: RequestHandler = async (context) => {
  const response = await handler(context)
  response.headers.set('Cross-Origin-Opener-Policy', 'same-origin')
  response.headers.set('Cross-Origin-Resource-Policy', 'cross-origin')

  return response
}

export default withHeaders
```

**Reply by paularmstrong** [reply]
This is a great step, but doesn't solve the problem of adding a content-security policy that uses `nonce` values. It feels like a glaring omission in the framework to not have both basic security headers and nonces easy to configure.

I've dug around a bunch and it doesn't seem like it's possible in `*-start` to add `nonce` attributes to the generated `<script>` tags.

**Reply by Enalmada** [reply]
true.  It will be impossible for any security conscious company to adopt tanstack start in production until a nonce can be passed through to the `<script>` tags.

**Comment by tannerlinsley** [comment]
This will be added in the next week-ish as we prepare for RC

**Reply by paularmstrong** [reply]
Amazing! Any issue or PR available to watch?

**Reply by guru-momos** [reply]
Hey, please let us know when a PR is available to watch for! 😄

**Reply by tannerlinsley** [reply]
We just merged this capability into the unified-route-tree-unsafe branch. We’ll be merging down into alpha asap.

**Reply by graup** [reply]
@tannerlinsley Any link to the PR / docs / examples? If you add that we could probably close this discussion.

**Reply by chrstntdd** [reply]
Found the current way to support setting global headers in Start: https://github.com/TanStack/router/issues/5407#issuecomment-3446993894

That gets part of the way there for a strict nonce-based CSP. What I'm still unclear on is how to get the server-side nonce set in the global middleware.

I see there was some recent work to allow passing nonce: https://github.com/TanStack/router/pull/5522

Not sure how to get that global nonce to `createRouter`:

```ts
export const getRouter = () => {
  return createRouter({
    routeTree,
    scrollRestoration: true,
    defaultPreloadStaleTime: 0,
     ssr: {
       nonce: "$global-nonce-from-server",
     },
  });
};
```

Seems like `getRouter` runs on server and client, too.

**Reply by hxtk** [reply]
I've arrived at the following implementation that sets a CSP header and nonce in global middleware and then uses the `startContext` on the server side to extract the nonce in `getRouter` and uses the meta tag produced by the server to extract the nonce on the client.

In `start.ts`:

```typescript
import { createMiddleware, createStart } from '@tanstack/react-start'
import {
  getResponseHeaders,
  setResponseHeaders,
  setResponseStatus,
} from '@tanstack/react-start/server'

const requestMiddleware = createMiddleware().server(({ next, context }) => {
  const nonce = Buffer.from(crypto.randomUUID()).toString('base64')
  const cspTemplate = `
    object-src 'none';
    script-src 'nonce-${nonce}' 'unsafe-inline' 'strict-dynamic' https: http:;
    style-src 'self' 'unsafe-inline';
    base-uri 'self';
    form-action 'self';
    frame-ancestors 'none';
    upgrade-insecure-requests;
  `

  // Replace newline characters and spaces
  const contentSecurityPolicyHeaderValue = cspTemplate
    .replace(/\s{2,}/g, ' ')
    .trim()

  const headers = getResponseHeaders()
  headers.set('Content-Security-Policy', contentSecurityPolicyHeaderValue)
  setResponseHeaders(headers)
  return next({
    context: {
      nonce,
    },
  })
})

export const startInstance = createStart(() => {
  return {
    requestMiddleware: [requestMiddleware],
  }
})
```

in `router.tsx`:

```typescript
import { createIsomorphicFn } from '@tanstack/react-start'
import { getStartContext } from '@tanstack/start-storage-context'
import { createRouter } from '@tanstack/react-router'
import { setupRouterSsrQueryIntegration } from '@tanstack/react-router-ssr-query'

// Import the generated route tree
import { routeTree } from './routeTree.gen'

const getNonce = createIsomorphicFn()
  .server(() => getStartContext().contextAfterGlobalMiddlewares.nonce)
  .client(() => document.querySelector("meta[property=csp-nonce]").content)

// Create a new router instance
export const getRouter = () => {
  const nonce = getNonce()

  const router = createRouter({
    routeTree,
    defaultPreload: 'intent',
    ssr: {
      nonce: nonce,
    },
  })

  setupRouterSsrQueryIntegration({ router, queryClient: rqContext.queryClient })

  return router
}
```

**Reply by chrstntdd** [reply]
@hxtk Thank you for sharing this!
I can repro on my side as well.

Would be nice to get strong types on the server side access of the middleware.

Also was not expecting to have to install a separate dependency (`@tanstack/start-storage-context`)  to access the global context.

Wondering if `getGlobalStartContext` in `@tanstack/react-start` could/would be the final place to read from?

--- 
edit: Oh, actually, this does work with types!

```ts
import { createRouter } from "@tanstack/react-router";

// Import the generated route tree
import { routeTree } from "./routeTree.gen";
import {
  createIsomorphicFn,
  getGlobalStartContext,
} from "@tanstack/react-start";

const getCspNonce = createIsomorphicFn()
  .server(() => {
    /**
     * @see https://github.com/TanStack/router/discussions/3028#discussioncomment-14844427
     * For reference
     */
    let ctx = getGlobalStartContext()!;
    let cspNonce = ctx.cspNonce;

    return cspNonce as string;
  })
  .client(() => {
    let el = document.querySelector(
      "meta[property=csp-nonce]",
    ) as HTMLMetaElement;

    return el.content;
  });

// Create a new router instance
export const getRouter = () => {
  return createRouter({
    routeTree,
    scrollRestoration: true,
    defaultPreloadStaleTime: 0,
    ssr: {
      nonce: getCspNonce(),
    },
  });
};

```

**Reply by John-Hejzlar** [reply]
Does this still work? I am getting this error unfortunately :/

```
39 | }): TThrow extends false
40 |   ? StartStorageContext | undefined
41 |   : StartStorageContext {
42 |   const context = startStorage.getStore()
43 |   if (!context && opts?.throwIfNotFound !== false) {
44 |     throw new Error(
         ^
error: No Start context found in AsyncLocalStorage. Make sure you are using the function within the server runtime.
```

**Comment by Enalmada** [comment]
I made this package to be able to share header and csp security setup between lots of projects:
https://github.com/Enalmada/start-secure

Implementation example:
https://github.com/Enalmada/tanstarter

**Comment by MagneH** [comment]
Got it working in my repo. I created a start.ts file in root and added a global middleware there. Picked it up in the router and added it in the router's config object.

Full example here: https://github.com/MagneH/SanTan

**Comment by Bazze** [comment]
It would be nice to be able to reuse existing libraries like helmet for this – anything that would be possible? [Helmet without Express](https://helmetjs.github.io/faq/use-without-express/) is possible, but I can't see that tanstack start is giving us access to the response object (I'm new to TanStack Start).

**Reply by CanRau** [reply]
You can modify response headers using `setResponseHeaders` or in middleware you can `await next()` which is basically the response, like shown here <https://tanstack.com/start/latest/docs/framework/react/guide/middleware#progressing-the-middleware-chain>

**Comment by CanRau** [comment]
Was really confused by `nonce` attribute applied to most head elements but without any value, which made me question my setup, until I found this https://stackoverflow.com/a/55673767/3484824 explaining why it's desired/expected behavior, in any case you can confirm the nonce is in the html by inspecting the page source, or from the network tab, the main document type request's "response" tab
